crates/code_generation/src/execution.rs

```diff
@@ -166,6 +285,91 @@ pub fn generate_replayer_struct(cyclers: &Cyclers) -> TokenStream {
                     _ => bail!("unexpected cycler instance name {cycler_instance_name}"),
                 }
             }
+
+            #(#reader_accessors)*
+        }
+    }
+}
+
+fn generate_replayer_fields_and_accessors(
+    cyclers: &Cyclers,
+    with_communication: bool,
+) -> (TokenStream, TokenStream, TokenStream) {
+    if with_communication {
+        (
+            quote! {
+                _communication_server: communication::server::Runtime<crate::structs::Parameters>,
+            },
+            quote! {
+                _communication_server: communication_server,
+            },
+            Default::default(),
+        )
+    } else {
+        let reader_tokens: Vec<_> = cyclers
+            .instances()
+            .map(|(cycler, instance)| {
+                (
+                    format_ident!("{}_reader", instance.to_case(Case::Snake)),
+                    format_ident!("{}", cycler.name.to_case(Case::Snake)),
+                )
+            })
+            .collect();
+
+        let reader_identifiers = reader_tokens.iter().map(|(reader, _cycler)| reader);
+        let reader_fields = reader_tokens.iter().map(|(reader, cycler)| {
+            quote! {
+                #reader: framework::Reader<crate::cyclers::#cycler::Database>,
+            }
+        });
+        let reader_accessors = reader_tokens.iter().map(|(reader, cycler)| {
+            quote! {
+                #[allow(unused)]
+                pub(crate) fn #reader(&self) -> framework::Reader<crate::cyclers::#cycler::Database> {
```

Comment:
1. I used `pub(crate)` since Database is also `pub(crate)` and would otherwise be more private than the method. Should we instead make Database `pub` as well?
2. This follows from 1. since there is a warning when the method is not used in the same crate. Though this could be removed in this case, since both `vision_top_reader()` and `vision_bottom_reader()` are used for imagine.

Reply:
Ok, sounds good, let's keep it as is.